lightning/src/util/sweep.rs

```diff
@@ -0,0 +1,688 @@
+	/// [`DEFAULT_REGENERATE_SPEND_THRESHOLD`] will be used.
+	///
+	/// The given `spend_outputs_callback` is a function takes a list of
+	/// [`SpendableOutputDescriptor`] and returns a fully signed ready-for-broadcast
+	/// [`Transaction`]. Usually, this should retrieve a change address from the on-chain wallet
+	/// and call [`KeysManager::spend_spendable_outputs`].
+	///
+	/// [`KeysManager::spend_spendable_outputs`]: crate::sign::KeysManager::spend_spendable_outputs
+	///
+	/// #### Example:
+	/// ```
+	/// # use bitcoin::key::Secp256k1;
+	/// # use bitcoin::{Network, Script, ScriptBuf, Transaction, Txid};
+	/// # use std::sync::Arc;
+	/// # use lightning::sign::SpendableOutputDescriptor;
+	/// # use lightning::sign::KeysManager;
+	/// # use lightning::chain::chaininterface::{ConfirmationTarget, FeeEstimator, BroadcasterInterface};
+	/// # use lightning::chain::{BestBlock, Filter, WatchedOutput};
+	/// # use lightning::util::sweep::OutputSweeper;
+	/// # use lightning::util::logger::{Logger, Record};
+	/// # use lightning::io;
+	/// # struct MyStore {}
+	/// # impl lightning::util::persist::KVStore for MyStore {
+	/// #     fn read(&self, primary_namespace: &str, secondary_namespace: &str, key: &str) -> io::Result<Vec<u8>> { Ok(Vec::new()) }
+	/// #     fn write(&self, primary_namespace: &str, secondary_namespace: &str, key: &str, buf: &[u8]) -> io::Result<()> { Ok(()) }
+	/// #     fn remove(&self, primary_namespace: &str, secondary_namespace: &str, key: &str, lazy: bool) -> io::Result<()> { Ok(()) }
+	/// #     fn list(&self, primary_namespace: &str, secondary_namespace: &str) -> io::Result<Vec<String>> { Ok(Vec::new()) }
+	/// # }
+	/// # struct MyWallet {}
+	/// # impl MyWallet {
+	/// #     fn get_new_address(&self) -> ScriptBuf { ScriptBuf::new() }
+	/// # }
+	/// # struct MyFeeEstimator {}
+	/// # impl FeeEstimator for MyFeeEstimator {
+	/// #     fn get_est_sat_per_1000_weight(&self, confirmation_target: ConfirmationTarget) -> u32 { 0 }
+	/// # }
+	/// # struct MyBroadcaster {}
+	/// # impl BroadcasterInterface for MyBroadcaster {
+	/// #     fn broadcast_transactions(&self, txs: &[&Transaction]) {}
+	/// # }
+	/// # struct MyFilter {}
+	/// # impl Filter for MyFilter {
+	/// #     fn register_tx(&self, txid: &Txid, script_pubkey: &Script) {}
+	/// #     fn register_output(&self, output: WatchedOutput) {}
+	/// # }
+	/// # struct MyLogger {}
+	/// # impl Logger for MyLogger {
+	/// #     fn log(&self, record: Record) {}
+	/// # }
+	/// # let my_wallet = Arc::new(MyWallet{});
+	/// # let my_fee_estimator = Arc::new(MyFeeEstimator{});
+	/// # let my_broadcaster = Arc::new(MyBroadcaster {});
+	/// # let my_keys_manager = Arc::new(KeysManager::new(&[42u8; 32], 0, 0));
+	/// # let my_store = Arc::new(MyStore {});
+	/// # let my_logger = Arc::new(MyLogger {});
+	/// # let my_best_block = BestBlock::from_network(Network::Regtest);
+	/// # let my_chain_data_source = Arc::new(MyFilter {});
```

Comment:
You can reduce this quite a bit by wrapping the code below in a function that takes the trait types as `dyn` parameters. Likewise for other variables here to avoid needing to initialize anything.

Reply:
Here's one using type aliases:

https://github.com/lightningdevkit/rust-lightning/blob/7ab438ddcc4eeccf84a4c80d982f0a3c025f2992/lightning-background-processor/src/lib.rs#L559-L572

You actually don't need `dyn` if you use type parameters on the function like the following.

https://github.com/jkczyz/rust-lightning/blob/956229dc5ae98a168636a2f30d97949cf49d1b0a/lightning/src/ln/channelmanager.rs#L1244

There I probably should use a trait for `Wallet`. The important part is using a function. It's basically a more general case of this, extending the idea to parameters:

https://doc.rust-lang.org/rustdoc/write-documentation/documentation-tests.html#using--in-doc-tests

Only if the function isn't `main`, then it's not run, only compiled, IIUC.

Reply:
Ah, I was only referring to the doctest just to save a few lines with defining an `unimplemented!` method body on a struct. `Wallet`'s definition and use as a method parameter would be hidden behind `#`. But having a non-doctest trait would indeed remove needing a definition in the doctest entirely.